x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/support/WatcherIndexTemplateRegistry.java

```diff
@@ -5,240 +5,81 @@
-            addIndexLifecyclePolicyIfMissing(state);
-        }
-    }
-
-    private void addTemplatesIfMissing(ClusterState state) {
-        boolean ilmSupported = XPackSettings.INDEX_LIFECYCLE_ENABLED.get(this.nodeSettings);
-        final TemplateConfig[] indexTemplates = ilmSupported ? TEMPLATE_CONFIGS : TEMPLATE_CONFIGS_NO_ILM;
-        for (TemplateConfig template : indexTemplates) {
-            final String templateName = template.getTemplateName();
-            final AtomicBoolean creationCheck = templateCreationsInProgress.computeIfAbsent(templateName, key -> new AtomicBoolean(false));
-            if (creationCheck.compareAndSet(false, true)) {
-                if (!state.metaData().getTemplates().containsKey(templateName)) {
-                    logger.debug("adding index template [{}], because it doesn't exist", templateName);
-                    putTemplate(template, creationCheck);
-                } else {
-                    creationCheck.set(false);
-                    logger.trace("not adding index template [{}], because it already exists", templateName);
-                }
-            }
-        }
-    }
-
-    private void putTemplate(final TemplateConfig config, final AtomicBoolean creationCheck) {
-        final Executor executor = threadPool.generic();
-        executor.execute(() -> {
-            final String templateName = config.getTemplateName();
-
-            PutIndexTemplateRequest request = new PutIndexTemplateRequest(templateName).source(config.load(), XContentType.JSON);
-            request.masterNodeTimeout(TimeValue.timeValueMinutes(1));
-            executeAsyncWithOrigin(client.threadPool().getThreadContext(), WATCHER_ORIGIN, request,
-                    new ActionListener<AcknowledgedResponse>() {
-                        @Override
-                        public void onResponse(AcknowledgedResponse response) {
-                            creationCheck.set(false);
-                            if (response.isAcknowledged() == false) {
-                                logger.error("Error adding watcher template [{}], request was not acknowledged", templateName);
-                            }
-                        }
-
-                        @Override
-                        public void onFailure(Exception e) {
-                            creationCheck.set(false);
-                            logger.error(new ParameterizedMessage("Error adding watcher template [{}]", templateName), e);
-                        }
-                    }, client.admin().indices()::putTemplate);
-        });
-    }
-
-    // Package visible for testing
-    LifecyclePolicy loadWatcherHistoryPolicy() {
-        return LifecyclePolicyUtils.loadPolicy(POLICY_WATCH_HISTORY.policyName, POLICY_WATCH_HISTORY.fileName, xContentRegistry);
+    protected List<IndexTemplateConfig> getTemplateConfigs(boolean ilmEnabled) {
+        return ilmEnabled ? TEMPLATE_CONFIGS : TEMPLATE_CONFIGS_NO_ILM;
```

Comment:
Do we need these constants? It seems like we can put all of the functionality of returning the proper templates in this method. Also, is the ilmEnabled (backed by the property) something thats dynamic? If not, we can probably put this all in the constructor and just hold a list of the proper templates as a member variable too.

Reply:
We don't need these constants - I've removed them.

The `ilmEnabled` boolean (as backed by the property) is not dynamic, so we could remove that parameter and just do that work in the constructor. I haven't yet, but may still do that but I wanted to get the rest of the changes up.

Reply:
After sleeping on it, I like this change and I've made it just set up the templates to use in the constructor.

This method could take the cluster state as a parameter, but I'm going to leave for if/when it becomes necessary.